walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,83 @@
 
 package com.google.sps;
 
+import java.util.ArrayList;
 import java.util.Collection;
+import java.util.List;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
 
 public final class FindMeetingQuery {
+  /**
+   * Finds all the time ranges in which the requested meeting can take place.
+   *
+   * @param events: the events that are already occuring.
+   * @param request: the request for the new event that we want to create.
+   * @return the possible time ranges for the request.
+   */
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    long duration = request.getDuration();
+    Collection<String> allAttendees =
+        Stream.concat(request.getAttendees().stream(), request.getOptionalAttendees().stream())
+            .collect(Collectors.toList());
+    Collection<TimeRange> possibleRangesWithOptional = getMeetingRanges(events, allAttendees, duration);
+    return (possibleRangesWithOptional.isEmpty())
+        ? getMeetingRanges(events, request.getAttendees(), duration)
+        : possibleRangesWithOptional;
+  }
+
+  /**
+   * Finds all the time ranges in which the meeting can be held.
+   * @param events: all the existing events that we have to consider.
+   * @param attendees: the attendees of the new event that we want to create.
+   * @param duration: the duration of the new event that we want to create.
+   * @return the time ranges that are free to use.
+   */
+  private Collection<TimeRange> getMeetingRanges(Collection<Event> events,
+                                                 Collection<String> attendees, long duration) {
+    List<TimeRange> busyTimeRanges = getBusyTimeRanges(events, attendees);
+    return getFreeRanges(busyTimeRanges, duration);
+  }
+
+  /**
+   * Finds all the time ranges that aren't busy and last at least as long as the given duration.
+   * @param busyTimeRanges: the time ranges that already have relevant events.
+   * @param duration: the duration of the new event that we want to create.
+   * @return the time ranges that are free to use.
+   */
+  private Collection<TimeRange> getFreeRanges(List<TimeRange> busyTimeRanges, long duration) {
+    Collection<TimeRange> freeTimeBlocks = new ArrayList<>();
+    int nextFreeStartTime = TimeRange.START_OF_DAY;
+    for (TimeRange currBusyTime : busyTimeRanges) {
+      int currBusyStartTime = currBusyTime.start();
+      if (currBusyStartTime - nextFreeStartTime >= duration) {
+        freeTimeBlocks.add(TimeRange.fromStartEnd(nextFreeStartTime, currBusyStartTime, false));
+      }
+      nextFreeStartTime = Math.max(currBusyTime.end(), nextFreeStartTime);
+    }
+    if (TimeRange.END_OF_DAY - nextFreeStartTime >= duration) {
+      freeTimeBlocks.add(TimeRange.fromStartEnd(nextFreeStartTime, TimeRange.END_OF_DAY, true));
```

Comment:
It's recommended to add a comment on non-named function arguments that states their name. 

For instance, here that'd be: 
`TimeRange.fromStartEnd(nextFreeStartTime, TimeRange.END_OF_DAY, /*inclusive*/ true))`

Reply:
Great, thanks for introducing this concept. Done.